src/DomainHost.sol

```diff
@@ -0,0 +1,411 @@
+    uint256 public line;        // Remote domain global debt ceiling [RAD]
+    uint256 public grain;       // Keep track of the pre-minted DAI in the escrow [WAD]
+    uint256 public sin;         // A running total of how much is required to re-capitalize the remote domain [WAD]
+    uint256 public cure;        // The amount of unused debt [RAD]
+    bool public cureReported;   // Returns true if cure has been reported by the guest
+    uint256 public live;
+
+    uint256 constant RAY = 10 ** 27;
+
+    string constant ARITHMETIC_ERROR = string(abi.encodeWithSignature("Panic(uint256)", 0x11));
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event Lift(uint256 wad);
+    event Release(uint256 wad);
+    event Push(int256 wad);
+    event Rectify(uint256 wad);
+    event Cage();
+    event Tell(uint256 value);
+    event Exit(address indexed usr, uint256 wad, uint256 claim);
+    event UndoExit(address indexed originalSender, uint256 wad);
+    event Deposit(address indexed to, uint256 amount);
+    event UndoDeposit(address indexed originalSender, uint256 amount);
+    event Withdraw(address indexed to, uint256 amount);
+    event RegisterMint(TeleportGUID teleport);
+    event InitializeRegisterMint(TeleportGUID teleport);
+    event FinalizeRegisterMint(TeleportGUID teleport);
+    event Settle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount);
+    event InitializeSettle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount);
+    event FinalizeSettle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount);
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "DomainHost/not-authorized");
+        _;
+    }
+
+    function _isGuest(address usr) internal virtual view returns (bool);
+    modifier guestOnly {
+        require(_isGuest(msg.sender), "DomainHost/not-guest");
+        _;
+    }
+
+    modifier vatLive {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+        _;
+    }
+
+    modifier ordered(uint256 _lid) {
+        require(lid++ == _lid, "DomainHost/out-of-order");
+        _;
+    }
+
+    constructor(bytes32 _ilk, address _daiJoin, address _escrow, address _router) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+        
+        ilk = _ilk;
+        daiJoin = DaiJoinLike(_daiJoin);
+        vat = daiJoin.vat();
+        dai = daiJoin.dai();
+        escrow = _escrow;
+        router = RouterLike(_router);
+        
+        vat.hope(_daiJoin);
+        dai.approve(_daiJoin, type(uint256).max);
+        dai.approve(_router, type(uint256).max);
+
+        live = 1;
+    }
+
+    // --- Math ---
+    function _int256(uint256 x) internal pure returns (int256 y) {
+        require((y = int256(x)) >= 0, ARITHMETIC_ERROR);
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x + y - 1) / y;
```

Comment:
nit: we should double check, but I think this was the most gas efficient way we had agreed to do it from now on:
```suggestion
            unchecked{
                z = x != 0 ? ((x - 1) / y) + 1 : 0;
            }
```

Reply:
Yes this is the better one.